### PR TITLE
Remove deprecated classes (ChannelIndex, Unit) and IOs; simplify tests

### PR DESCRIPTION
**Pull request by apdavison**
This replaces #943.

- removes the deprecated core classes (ChannelIndex, Unit)
- simplifies the unit tests for core classes

Please review carefully, especially the changes to IO modules.

It does not include removing use of ChannelIndex, Unit from the examples, that still needs to be done.

### CODE REVIEW

File: neo/core/segment.py [hunk 1]
```diff
@@ -139,129 +139,6 @@ def t_stop(self):
...
-        '''
-        # todo: provide equivalent method using Group/ChannelView
-        #       add deprecation message (use decorator)?
-        seg = Segment()
-        seg.spiketrains = self.take_spiketrains_by_unit(unit_list)
-        seg.analogsignals = \
-            self.take_slice_of_analogsignalarray_by_unit(unit_list)
-        # TODO copy others attributes
-        return seg
-
```
Comment (samuelgarcia) [inline neo/core/segment.py:264]: yeag

**Comment by apdavison** [comment]
@JuliaSprenger @samuelgarcia please could you take a look at this? I'm fairly sure the failing tests are unrelated to this PR.

I'm waiting for this to be merged before I can make the SpikeTrainList PR.

File: neo/test/generate_datasets.py [hunk 2]
```diff
@@ -18,6 +19,216 @@
...
+    block.segments[0].events.extend((
+        random_event(name="event array #1 in segment #1", thing="frooble"),
+    ))
+    block.segments[1].events.extend((
+        random_event(name="event array #1 in segment #2", thing="wotsit"),
+    ))
+    block.segments[0].spiketrains.extend((
+        random_spiketrain(name="spiketrain #1 in segment #1", thing="frooble"),
+        random_spiketrain(name="spiketrain #2 in segment #1", thing="wotsit")
+    ))
```
Comment (JuliaSprenger) [inline neo/test/generate_datasets.py:228]:
```suggestion
        random_signal(name="signal #2 in segment #1", thing="frooble")))
        
    block.segments[1].analogsignals.extend((
        random_signal(name="signal #1 in segment #2", thing="amajig")))
        
    block.segments[0].events.extend((
        random_event(name="event array #1 in segment #1", thing="frooble")))
        
    block.segments[1].events.extend((
        random_event(name="event array #1 in segment #2", thing="wotsit")))
        
    block.segments[0].spiketrains.extend((
        random_spiketrain(name="spiketrain #1 in segment #1", thing="frooble"),
        random_spiketrain(name="spiketrain #2 in segment #1", thing="wotsit")))
```

File: neo/test/iotest/test_nixio.py [hunk 3]
```diff
@@ -67,11 +67,6 @@ def compare_blocks(self, neoblocks, nixblocks):
...
                 nixgrp = nixblock.groups[neoseg.annotations["nix_name"]]
                 self.compare_segment_group(neoseg, nixgrp)
-            self.assertEqual(len(neoblock.channel_indexes),
-                             len(nixblock.sources))
-            for idx, neochx in enumerate(neoblock.channel_indexes):
-                nixsrc = nixblock.sources[neochx.annotations["nix_name"]]
-                self.compare_chx_source(neochx, nixsrc)
             self.check_refs(neoblock, nixblock)
 
     def compare_chx_source(self, neochx, nixsrc):
```
Comment (JuliaSprenger) [inline neo/test/iotest/test_nixio.py:72]: this test method also still uses channelindex references

**Review by JuliaSprenger — COMMENTED** [review]
Thanks @apdavison for the thorough cleanup. I undestand that splitting this PR into two dedicated PRs one for the removal of deprecated classes and a second one for the simplification of tests would be quite complex and require a lot of rewrite. If @samuelgarcia agrees we can leave these two topics combined in this PR.

**Comment by JuliaSprenger** [comment]
It seems the failing tests are related to disk space, maybe circleci provides less space since the last test runs? In any case we might need to start cleaning test files after using them...

File: neo/io/nixio.py [hunk 4]
```diff
@@ -331,13 +331,6 @@ def _nix_to_neo_block(self, nix_block):
             if mt.type == "neo.spiketrain" and mt.name not in self._neo_map:
                 self._nix_to_neo_spiketrain(mt)
 
-        # descend into Sources
```
[the remark this answers is not in the document]
Comment (apdavison) [reply neo/io/nixio.py:334]: I thought we agreed that to read such files people should install Neo 0.9, read and re-save the files, then install a more recent version of Neo. The reason is that there is no single, guaranteed correct way to map ChannelIndex/Unit to Group

**Comment by samuelgarcia** [comment]
Hi Andrew.
This big clean is amazing!!
This is OK for me.
neo object generation simplified is really great.

Also `filter()` testing which have been lightened is also ok for me because this filter have to be enterily redesigned.

File: neo/test/coretest/test_block.py [hunk 5]
```diff
@@ -802,21 +440,7 @@ def test__deepcopy(self):
...
-            for sig in chidx.irregularlysampledsignals:
-                self.assertEqual(id(sig.channel_index), id(chidx))
-
-            for unit in chidx.units:
-                self.assertEqual(id(unit.channel_index), id(chidx))
-
-                for sptr in unit.spiketrains:
-                    self.assertEqual(id(sptr.unit), id(unit))
-
+        # todo: check all other block contents, especially Groups
```
[the remark this answers is not in the document]
Comment (apdavison) [reply neo/test/coretest/test_block.py:443]: on looking at this again, `assert_same_sub_schema()` includes checks of all object attributes in the tree, so this "todo" is not necessary

**Comment by apdavison** [comment]
I think this is ready to merge - I've implemented the "todo" items, the tests pass, there are no conflicts.

**Comment by apdavison** [comment]
(there's one minor pep8 issue which I'll fix in some other PR)

**Comment by samuelgarcia** [comment]
OK for me.
@JuliaSprenger : I let you merge.

**Comment by apdavison** [comment]
please can this be merged soon so I can start working on the SpikeTrainList PR again?

**Comment by apdavison** [comment]
thanks, that was quick!

**Comment by samuelgarcia** [comment]
j'ai de bon reflex

**Pull request merged**